src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DCTree.java

```diff
@@ -53,41 +84,150 @@
     }
 
     /**
-     * {@return the source position for position relative to this tree node}
-     * This is primarily useful for nodes that wrap a single string child.
+     * {@return the start position of this tree node}
+     *
+     * For most nodes, this is the position of the first character that is unique to this node.
      *
-     * @param dc     the enclosing doc comment
-     * @param offset the offset
+     * The value is relative to the beginning of the comment text.
+     * Use {@link DCDocComment#getSourcePosition(int)} to convert
+     * it to a position in the source file.
      */
-    public long getSourcePosition(DCDocComment dc, int offset) {
-        return dc.comment.getSourcePos(pos + offset);
+    public int getStartPosition() {
+        return pos;
     }
 
-    public JCDiagnostic.DiagnosticPosition pos(DCDocComment dc) {
-        return new SimpleDiagnosticPosition(dc.comment.getSourcePos(pos));
+    /**
+     * {@return the "preferred" position of this tree node}
+     *
+     * It is typically the position of the first character that is unique to this node.
+     * It is the position that is used for the caret in "line and caret" diagnostic messages.
+     *
+     * The value is relative to the beginning of the comment text.
+     * Use {@link DCDocComment#getSourcePosition(int)} to convert
+     * it to a position in the source file.
+     */
+    public int getPreferredPosition() {
+        return pos;
     }
 
-    /** Convert a tree to a pretty-printed string. */
+    /**
+     * {@return the end position of the tree node}
+     *
+     * The value is typically derived in one of three ways:
+     * <ul>
+     * <li>computed from the start and length of "leaf" nodes, such as {@link TextTree},
+     * <li>computed recursively from the end of the last child node, such as for most {@link DCBlockTag block tags}, or
+     * <li>provided explicitly, such as for subtypes of {@link DCEndPosTree}
+     * </ul>
+     *
+     * The value is relative to the beginning of the comment text.
+     * Use {@link DCDocComment#getSourcePosition(int)} to convert
+     * it to a position in the source file.
+     */
+    public int getEndPosition() {
+        if (this instanceof DCEndPosTree<?> dcEndPosTree) {
+            int endPos = dcEndPosTree.getEndPos();
+
+            if (endPos != NOPOS) {
+                return endPos;
+            }
+        }
+
+        switch (getKind()) {
+            case TEXT -> {
+                DCText text = (DCText) this;
+                return text.pos + text.text.length();
+            }
+
+            case ERRONEOUS -> {
+                DCErroneous err = (DCErroneous) this;
+                return err.pos + err.body.length();
+            }
+
+            case IDENTIFIER -> {
+                DCIdentifier ident = (DCIdentifier) this;
+                return ident.pos + ident.name.length();
```

Comment:
While in JavacTrees, this logic contained some special handling for `names.error`. What was that about?

Reply:
`names.error` is used by javac in various error conditions. It has the string  value `<error>`.  I don't believe it can arise in `DocTree` nodes, so I dropped the code.